unitary/alpha/quantum_world.py

```diff
@@ -689,26 +690,103 @@ def density_matrix(
+            )
+            result += cirq.von_neumann_entropy(partial_density_matrix, validate=False)
+        result -= cirq.von_neumann_entropy(density_matrix, validate=False)
+        return result
+
+    def print_entanglement_table(self, count: int = 1000) -> None:
+        """Peek the current quantum world `count` times, and calculate pair-wise entanglement
+        (i.e. quantum mutual information) for each pair of quantum objects.
+        See https://en.wikipedia.org/wiki/Quantum_mutual_information for the formula. And print
+        the results out in a table.
+
+        Parameters:
+            count:      Number of measurements.
+        """
+        objects = list(self.object_name_dict.values())
+        num_qubits = len(objects)
+        if num_qubits < 2:
+            raise ValueError(
+                f"There is only {num_qubits} qubit in the quantum world. "
+                "At least 2 qubits are required to calculate entanglements."
+            )
+        # Peek the current world `count` times and get the results.
+        histogram = self.get_correlated_histogram(objects, count)
+
+        # Get an estimate of the state vector.
+        state_vector = np.array([0.0] * (2**num_qubits))
+        for key, val in histogram.items():
+            state_vector += self.__to_state_vector__(key) * np.sqrt(val * 1.0 / count)
+        density_matrix = np.outer(state_vector, state_vector)
+        reshaped_density_matrix = density_matrix.reshape((2, 2) * num_qubits)
+
+        entropy = [0.0] * num_qubits
+        entropy_pair = np.zeros((num_qubits, num_qubits))
+        entanglement = np.zeros((num_qubits, num_qubits))
+        for i in range(num_qubits - 1):
+            for j in range(i + 1, num_qubits):
+                density_matrix_ij = cirq.partial_trace(reshaped_density_matrix, [i, j])
+                entropy_pair[i][j] = cirq.von_neumann_entropy(
+                    density_matrix_ij.reshape(4, 4), validate=False
+                )
+                if i == 0:
+                    # Fill in entropy [0]
+                    if j == i + 1:
+                        density_matrix_i = cirq.partial_trace(density_matrix_ij, [0])
+                        entropy[i] = cirq.von_neumann_entropy(
+                            density_matrix_i, validate=False
+                        )
+                    # Fill in entropy [1 to num_qubit - 1]
+                    density_matrix_j = cirq.partial_trace(density_matrix_ij, [1])
+                    entropy[j] = cirq.von_neumann_entropy(
+                        density_matrix_j, validate=False
+                    )
+                entanglement[i][j] = entropy[i] + entropy[j] - entropy_pair[i][j]
+                entanglement[j][i] = entanglement[i][j]
+        names = list(self.object_name_dict.keys())
+        data_frame = pd.DataFrame(entanglement, index=names, columns=names)
+        print(data_frame.round(1))
```

Comment:
Should we return the data_frame (or some other result) instead of printing it? 
Printing it seems generally unhelpful if we are using this within a game.